### PR TITLE
Visual Data Mapping support

### PR DESCRIPTION
**Pull request by vyeshwanth**
## Description of the change

>  disables transforming events to destination for events already mapped to destination. Events sent by warehouses could be mapped to destination already. Hence, no destination transformation is needed for such events. 
Flag used to check if the event is mapped to destination is `mappedToDestination` in the message

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Comment by ruchiramoitra** [comment]
LGTM @arnab-p  can you review too

File: v0/util/index.js [hunk 1]
```diff
@@ -953,6 +953,14 @@ function returnArrayOfSubarrays(arr, len) {
   return chunks;
 }
 
+// Helper method to add external Id to traits
+// Traverse through the possible keys for traits using generic mapping and add externalId if traits found
+function addExternalIdToTraits(message) {
+    const identifierType = message.context.externalId[0].identifierType;
```
Comment (arnab-p) [inline v0/util/index.js:959]: According to our previous schema, `externalId` objects have fields `type` and `id`. Are we changing this purposefully? If not, we should keep it same.

File: v0/util/index.js [hunk 2]
```diff
@@ -953,6 +953,14 @@ function returnArrayOfSubarrays(arr, len) {
...
+    const identifierValue = message.context.externalId[0].id;
+    set(getFieldValueFromMessage(message, "traits"), identifierType, identifierValue);
```
Comment (arnab-p) [inline v0/util/index.js:961]: Should we check the value for null/undefined/falsy values before setting it?

**Review by arnab-p — CHANGES_REQUESTED** [review]

File: v0/util/index.js (hunk 1 above)
Comment (vyeshwanth) [reply v0/util/index.js:959]: `type` will be the Object type where as `identifierType` is the actual unique field  name and the `id` contains the value for that unique field. So, we just add the field and the value to traits.

File: v0/destinations/braze/transform.js [hunk 3]
```diff
@@ -100,6 +100,10 @@ function getUserAttributesObject(message, mappingJson) {
   // get traits from message
   const traits = getFieldValueFromMessage(message, "traits");
 
+  if(message.mappedToDestination) {
```
[the remark this answers is not in the document]
Comment (vyeshwanth) [reply v0/destinations/braze/transform.js]: updated and wrapped the flag in `meta`. Updated the tests accordingly

File: v0/destinations/braze/transform.js [hunk 4]
```diff
@@ -141,6 +145,12 @@ function getUserAttributesObject(message, mappingJson) {
 }
 
 function processIdentify(message, destination) {
+  
+  // override userId with externalId in context(if present) and event is mapped to destination
+  if(message.mappedToDestination && message.context.externalId[0].id) {
```
[the remark this answers is not in the document]
Comment (vyeshwanth) [reply v0/destinations/braze/transform.js]: Updated

File: v0/destinations/mailchimp/transform.js [hunk 5]
```diff
@@ -14,9 +14,11 @@ const {
   getFieldValueFromMessage,
   getSuccessRespEvents,
   getErrorRespEvents,
-  CustomError
+  CustomError,
+  addExternalIdToTraits
 } = require("../../util");
 const logger = require("../../../logger");
+const set = require("set-value");
```
[the remark this answers is not in the document]
Comment (vyeshwanth) [reply v0/destinations/mailchimp/transform.js]: updated

File: v0/util/index.js (hunk 2 above)
Comment (vyeshwanth) [reply v0/util/index.js]: set method is already making checks for undefined, nulls etc., It should suffice in this case. Please let me know your opinion

**Review by arnab-p — APPROVED** [review]

**Pull request merged**